functions/Get-DbaFirewallRule.ps1

```diff
@@ -0,0 +1,235 @@
+
+        Returns all firewall rules on the computer SRV1 related to SQL Server.
+        The value "AllOnComputer" only uses the computer name part of SqlInstance.
+
+    #>
+    [CmdletBinding()]
+    param (
+        [parameter(Mandatory, ValueFromPipeline)]
+        [DbaInstanceParameter[]]$SqlInstance,
+        [PSCredential]$Credential,
+        [ValidateSet('Engine', 'Browser', 'AllOnComputer')]
+        [string[]]$Type,
+        [switch]$EnableException
+    )
+
+    begin {
+        $cmdScriptBlock = {
+            # This scriptblock will be processed by Invoke-Command2.
+            try {
+                if (-not (Get-Command -Name Get-NetFirewallRule -ErrorAction SilentlyContinue)) {
+                    throw 'The module NetSecurity with the command Get-NetFirewallRule is missing on the target computer, so Get-DbaFirewallRule is not supported.'
+                }
+                $successful = $true
+                $verbose = @( )
+                $rules = Get-NetFirewallRule -Group 'SQL Server' -WarningVariable warn -ErrorVariable err -WarningAction SilentlyContinue -ErrorAction SilentlyContinue
+                if ($warn.Count -gt 0) {
+                    $successful = $false
+                } else {
+                    # Change from an empty System.Collections.ArrayList to $null for better readability
+                    $warn = $null
+                }
+                if ($err.Count -gt 0) {
+                    if ($err.Count -eq 1 -and $err[0] -match 'No MSFT_NetFirewallRule objects found') {
+                        $verbose += "No objects found. Detailed error message: $($err[0])"
+                        $err = $null
+                    } else {
+                        $successful = $false
+                    }
+                } else {
+                    # Change from an empty System.Collections.ArrayList to $null for better readability
+                    $err = $null
+                }
+                if ($successful) {
+                    $verbose += "Get-NetFirewallRule was successful, we have $($rules.Count) rules."
+                    $rulesWithDetails = @( )
+                    foreach ($rule in $rules) {
+                        $rulesWithDetails += [PSCustomObject]@{
+                            DisplayName = $rule.DisplayName
+                            Name        = $rule.Name
+                            Protocol    = ($rule | Get-NetFirewallPortFilter).Protocol
+                            LocalPort   = ($rule | Get-NetFirewallPortFilter).LocalPort
+                            Program     = ($rule | Get-NetFirewallApplicationFilter).Program
+                            Rule        = $rule
+                        }
+                    }
+                }
+                [PSCustomObject]@{
+                    Successful = $successful
+                    Rules      = $rulesWithDetails
+                    Verbose    = $verbose
+                    Warning    = $warn
```

Comment:
This is a deviation from the way we usually handle exceptions. Is this because of Remoting?

Reply:
Yes. Still not sure if this is the best way, may be changed...

Reply:
But the idea is to not fail on the remote machine and get as much info back to then report and fail on the local machine.